lib/fizzy/parser_expr.cpp

```diff
@@ -57,13 +58,25 @@ parser_result<Code> parse_expr(const uint8_t* pos, const uint8_t* end, bool have
     // For a block/if/else instruction the value is the block/if/else's immediate offset.
     Stack<LabelPosition> label_positions;
 
+    const auto metrics_table = get_instruction_metrics_table();
+    int stack_height = 0;
+
     bool continue_parsing = true;
     while (continue_parsing)
     {
         if (pos == end)
             throw parser_error{"Unexpected EOF"};
 
-        const auto instr = static_cast<Instr>(*pos++);
+        const auto opcode = *pos++;
+        const auto& metrics = metrics_table[opcode];
+
+        if (stack_height < metrics.stack_height_required)
+            throw validation_error{"stack underflow"};
+
+        stack_height += metrics.stack_height_change;
+        code.max_stack_height = std::max(code.max_stack_height, stack_height);
```

Comment:
max height is used only in tests now?

Reply:
It is used in the next PR: #247. I will move it there only.